zoneinfo/src/parser.rs

```diff
@@ -0,0 +1,200 @@
+    fn context_parse<T: TryFromStr<LineParseContext>>(
+        &self,
+        ctx: &mut LineParseContext,
+    ) -> Result<T, <T as TryFromStr<LineParseContext>>::Error> {
+        T::try_from_str(self, ctx)
+    }
+}
+
+impl ContextParse for String {
+    fn context_parse<T: TryFromStr<LineParseContext>>(
+        &self,
+        ctx: &mut LineParseContext,
+    ) -> Result<T, <T as TryFromStr<LineParseContext>>::Error> {
+        T::try_from_str(self, ctx)
+    }
+}
+
+impl TryFromStr<LineParseContext> for String {
+    type Error = ZoneInfoParseError;
+    fn try_from_str(s: &str, _: &mut LineParseContext) -> Result<Self, Self::Error> {
+        Ok(s.parse::<String>().expect("parse String is infallible"))
+    }
+}
+
+impl TryFromStr<LineParseContext> for i8 {
+    type Error = ZoneInfoParseError;
+    fn try_from_str(s: &str, ctx: &mut LineParseContext) -> Result<Self, Self::Error> {
+        s.parse::<i8>()
+            .map_err(|e| ZoneInfoParseError::ParseIntError(ctx.line_number, e, ctx.span()))
+    }
+}
+
+impl TryFromStr<LineParseContext> for u8 {
+    type Error = ZoneInfoParseError;
+    fn try_from_str(s: &str, ctx: &mut LineParseContext) -> Result<Self, Self::Error> {
+        s.parse::<u8>()
+            .map_err(|e| ZoneInfoParseError::ParseIntError(ctx.line_number, e, ctx.span()))
+    }
+}
+
+impl TryFromStr<LineParseContext> for u16 {
+    type Error = ZoneInfoParseError;
+    fn try_from_str(s: &str, ctx: &mut LineParseContext) -> Result<Self, Self::Error> {
+        s.parse::<u16>()
+            .map_err(|e| ZoneInfoParseError::ParseIntError(ctx.line_number, e, ctx.span()))
+    }
+}
+
+impl TryFromStr<LineParseContext> for i32 {
+    type Error = ZoneInfoParseError;
+    fn try_from_str(s: &str, ctx: &mut LineParseContext) -> Result<Self, Self::Error> {
+        s.parse::<i32>()
+            .map_err(|e| ZoneInfoParseError::ParseIntError(ctx.line_number, e, ctx.span()))
+    }
+}
+
+pub(crate) fn next_split<'a>(
+    splits: &mut SplitWhitespace<'a>,
+    context: &LineParseContext,
+) -> Result<&'a str, ZoneInfoParseError> {
+    splits.next().ok_or(ZoneInfoParseError::UnexpectedEndOfLine(
+        context.line_number,
+        context.span(),
+    ))
+}
+
+pub(crate) fn remove_comments(line: &str) -> &str {
+    if let Some((cleaned, _comment)) = line.split_once("#") {
+        cleaned
+    } else {
+        line
+    }
+}
+
+#[non_exhaustive]
+pub struct ZoneInfoParser<'data> {
+    lines: Peekable<Lines<'data>>,
+}
+
+impl<'data> ZoneInfoParser<'data> {
+    /// Creates a parser from a `&str`
+    pub fn from_zoneinfo_str(source: &'data str) -> Self {
+        Self {
+            lines: source.lines().peekable(),
+        }
+    }
+
+    #[allow(clippy::while_let_on_iterator)]
+    pub fn parse(&mut self) -> Result<ZoneInfoCompiler, ZoneInfoParseError> {
+        let mut zoneinfo = ZoneInfoCompiler::default();
+        let mut context = LineParseContext::default();
+        while let Some(line) = self.lines.peek() {
```

Comment:
suggestion: Document why this is `while let` on a `Peekable` instead of a `for line in self.lines`. It took me some work to find where it was used for anything other than `next()` inside the loop. Maybe also move the `allow` directly next to the block.